### PR TITLE
[MPS] Implement MPSGenerator to enable manual random seeding

### LINKED ISSUE #84288: MPS: torch.manual_seed not working on metal (mps)  for  torch.randn
### 🐛 Describe the bug

Using torch.manual_seed to set a seed for torch.nrand usage not not work, seeding with the same number does not bring back consistent results for  example rung the noddy script

```
import torch

torch.manual_seed(999)
print (torch.randn(3, device='cpu'))
torch.manual_seed(999)
print (torch.randn(3, device='cpu'))
torch.manual_seed(999)
print (torch.randn(3, device='mps'))
torch.manual_seed(999)
print (torch.randn(3, device='mps'))
torch.manual_seed(999)
print (torch.randn(3, device='cpu'))
torch.manual_seed(999)
print (torch.randn(3, device='cpu'))

```

returns

```
tensor([-0.8379,  0.4564,  0.3481])
tensor([-0.8379,  0.4564,  0.3481])
tensor([ 0.2612,  1.2390, -1.1139], device='mps:0')
tensor([-1.4076,  1.3567, -0.4096], device='mps:0')
tensor([-0.8379,  0.4564,  0.3481])
tensor([-0.8379,  0.4564,  0.3481])
```

Note the two MPS lines return different values I'd expect them to be the same, ideally I'd like them to be the same as
the CPU values but appreciate that may not be possible.

BTW I've cut out the following ou of those results in case you were wondered where it was :-)

```
/Volumes/Sabrent Media/Documents/Source/Python/diffuse/lib/python3.10/site-packages/torch/_tensor_str.py:114: UserWarning: The operator 'aten::masked_select' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(

```

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0.dev20220828
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.22.4
Libc version: N/A

Python version: 3.10.4 (main, May 10 2022, 03:52:14) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-12.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.2
[pip3] pytorch-lightning==1.4.2
[pip3] torch==1.13.0.dev20220828
[pip3] torch-fidelity==0.3.0
[pip3] torchaudio==0.13.0.dev20220827
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.14.0.dev20220827
[conda] Could not collect
```

BTW had to fix the collection script to deal with spaces in path names :-)

cc @pbelevich @kulinseth @albanD

### PR DESCRIPTION
**Pull request by razarmehr**
This patch adds support for creating torch.Generator for MPS device, and enables its functions such as manual_seed, get_state, and set_state.
Fixes #84288 and #84516

### CODE REVIEW

File: aten/src/ATen/Context.h [hunk 1]
```diff
@@ -424,6 +426,13 @@ static inline void manual_seed(uint64_t seed) {
       }
     }
   }
+
+  if (hasMPS()) {
+    auto mps_gen = globalContext().defaultGenerator(DeviceType::MPS);
+    // See Note [Acquire lock when using random generators]
+    std::lock_guard<std::mutex> lock(mps_gen.mutex());
```
Comment (albanD) [inline aten/src/ATen/Context.h:433]: Why is this lock outside of the object? Why isn't that handled within mps_gen?

File: aten/src/ATen/mps/MPSGeneratorImpl.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+TORCH_API const Generator& getDefaultMPSGenerator();
+TORCH_API Generator createMPSGenerator(uint64_t seed_val = default_rng_seed_val);
+
+} // namespace detail
+} // namespace mps
+
+struct TORCH_API MPSGeneratorImpl : public c10::GeneratorImpl {
+  // Constructors
+  MPSGeneratorImpl(uint64_t seed_in = default_rng_seed_val);
```
Comment (albanD) [inline aten/src/ATen/mps/MPSGeneratorImpl.h:28]: Do you need the default value here?

File: aten/src/ATen/mps/MPSGeneratorImpl.mm [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  memcpy(this->state_data(), new_rng_state, states_size);
+}
+
+std::shared_ptr<MPSGeneratorImpl> MPSGeneratorImpl::clone() const {
+  return std::shared_ptr<MPSGeneratorImpl>(this->clone_impl());
+}
+
+MPSGeneratorImpl* MPSGeneratorImpl::clone_impl() const {
+  auto gen = new MPSGeneratorImpl(this->data_.seed);
+  gen->set_current_seed(this->data_.seed);
```
Comment (albanD) [inline aten/src/ATen/mps/MPSGeneratorImpl.mm:96]: If you already create it with the right seed, why do you need to reset it here?

File: torch/csrc/Generator.cpp [hunk 4]
```diff
@@ -52,12 +56,20 @@ static PyObject* THPGenerator_pynew(
   auto device = r.deviceWithDefault(0, at::Device(at::kCPU));
 
   THPGeneratorPtr self((THPGenerator*)type->tp_alloc(type, 0));
-#ifdef USE_CUDA
+#if defined(USE_CUDA) || defined(USE_MPS)
```
Comment (albanD) [inline torch/csrc/Generator.cpp:59]: This logic is a bit weird tbh.
Let's just do:
```
if (device.type() == at::kCPU) {
    self->cdata = make_generator<CPUGeneratorImpl>();
}
#ifdef USE_CUDA
  else if (device.type() == at::kCUDA) {
    self->cdata = make_generator<CUDAGeneratorImpl>(device.index());
  }
#endif
#if USE_MPS
  else if (device.type() == at::kMPS) {
    self->cdata = make_generator<MPSGeneratorImpl>();
  }
#endif
else {
  TORCH_CHECK(false, ...);
}
```

**Review by albanD — COMMENTED** [review]
Sounds good. Only small things

File: aten/src/ATen/Context.h (hunk 1 above)
Comment (razarmehr) [reply aten/src/ATen/Context.h:433]: The lock depends on a Mutex within mps_gen itself. The note `[Acquire lock when using random generators]` explains why it's designed this way globally. It's used to protect setting the seeds and other random params. 
CUDA is using the same logic in the lines above this code:
```
auto cuda_gen = globalContext().defaultGenerator(
          Device(at::kCUDA, static_cast<c10::DeviceIndex>(i)));
      {
        // See Note [Acquire lock when using random generators]
        std::lock_guard<std::mutex> lock(cuda_gen.mutex());
        cuda_gen.set_current_seed(seed);
      }  
```

File: aten/src/ATen/mps/MPSGeneratorImpl.h (hunk 2 above)
Comment (razarmehr) [reply aten/src/ATen/mps/MPSGeneratorImpl.h:28]: The code is based on the existing `CPUGeneratorImpl.h`:
```
struct TORCH_API CPUGeneratorImpl : public c10::GeneratorImpl {
  CPUGeneratorImpl(uint64_t seed_in = default_rng_seed_val);
...
createCPUGenerator(uint64_t seed_val = default_rng_seed_val);
```

It's just there to set a default value when we create a Generator's instance implicitly without calling the createCPUGenerator().

File: aten/src/ATen/mps/MPSGeneratorImpl.mm (hunk 3 above)
Comment (razarmehr) [reply aten/src/ATen/mps/MPSGeneratorImpl.mm:96]: The `set_current_seed()` does not only reset seeds. But it sets some other params for Philox engine too which are calculated based on seed:
```
void MPSGeneratorImpl::set_current_seed(uint64_t seed) {
  data_.seed = seed;
  data_.state.fill(1);
  // the two last state values are the Philox keys
  // TODO: make "key" in PhiloxRNGEngine.h public so we don't duplicate code here
  data_.state[5] = static_cast<uint32_t>(seed);
  data_.state[6] = static_cast<uint32_t>(seed >> 32);
  engine_.reset_state(seed);
}
```

File: torch/csrc/Generator.cpp (hunk 4 above)
Comment (razarmehr) [reply torch/csrc/Generator.cpp:59]: This was based on the original weird CUDA logic which creates "CPU" instance in a `#ifdef USE_CUDA` block:
```
#ifdef USE_CUDA
  if (device.type() == at::kCPU) {
    self->cdata = make_generator<CPUGeneratorImpl>();
  } else if (device.type() == at::kCUDA) { 
```
I didn't want to modify the original code, so I just married MPS into it :)
I'll try your suggestion.

File: aten/src/ATen/mps/MPSGeneratorImpl.mm (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/mps/MPSGeneratorImpl.mm:96]: So the MPSGeneratorImpl() could be created with the default value and doesn't need the seed?

File: torch/csrc/Generator.cpp (hunk 4 above)
Comment (albanD) [reply torch/csrc/Generator.cpp:59]: Feel free to cleanup weird cuda code ;)

Comment (razarmehr) [reply torch/csrc/Generator.cpp]: Updated the PR.

File: aten/src/ATen/mps/MPSGeneratorImpl.mm (hunk 3 above)
Comment (razarmehr) [reply aten/src/ATen/mps/MPSGeneratorImpl.mm:96]: Yes, I could've have let `set_current_seed()` to set the seed instead of passing via the constructor above it. 
I updated the PR.

File: aten/src/ATen/detail/MPSHooksInterface.h [hunk 5]
```diff
@@ -28,7 +28,7 @@ struct TORCH_API MPSHooksInterface {
     return false;
   }
 
-  virtual const Generator& getDefaultMPSGenerator(C10_UNUSED DeviceIndex device_index = -1) const {
+  virtual const Generator& getDefaultMPSGenerator() const {
```
Comment (malfet) [inline aten/src/ATen/detail/MPSHooksInterface.h:31]: Hmm, I think it should take the device index, but error out if index is other than `0` (as `index` is inherent property of the device and it's existence is not checked when device is created)

File: aten/src/ATen/Context.h [hunk 6]
```diff
@@ -38,6 +38,8 @@ class TORCH_API Context {
       return at::detail::getDefaultCPUGenerator();
     } else if (device_type == at::kCUDA) {
       return at::detail::getCUDAHooks().getDefaultCUDAGenerator(device.index());
+    } else if (device_type == at::kMPS) {
+      return at::detail::getMPSHooks().getDefaultMPSGenerator();
```
Comment (malfet) [inline aten/src/ATen/Context.h:42]: Let's pass the index anyway..
```suggestion
      return at::detail::getMPSHooks().getDefaultMPSGenerator(device.index());
```

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/detail/MPSHooksInterface.h (hunk 5 above)
Comment (razarmehr) [reply aten/src/ATen/detail/MPSHooksInterface.h:31]: I removed it because there was no use case for it. I can add it but it'll be used only in a check like `TORCH_CHECK(device_index == 0, ...)`. 
The other options is to add that later when we support multiple devices.

File: aten/src/ATen/Context.h (hunk 6 above)
Comment (razarmehr) [reply aten/src/ATen/Context.h:42]: Please see my comment below, and let me know if you'd think it's better to add it now or later when we support multiple MPS devices.

Comment (razarmehr) [reply aten/src/ATen/Context.h:42]: Adding device index down to MPSGenerator is a bigger change (it's not just `getDefaultMPSGenerator()`), and I'll do that in a separate PR.

**Review by albanD — APPROVED** [review]
SGTM
